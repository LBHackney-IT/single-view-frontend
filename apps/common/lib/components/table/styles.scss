@import "~lbh-frontend/lbh/base";
@import "~lbh-frontend/lbh/components/lbh-table/table";
