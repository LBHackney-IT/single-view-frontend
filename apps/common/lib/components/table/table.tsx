import React, { ComponentPropsWithoutRef, forwardRef } from "react";

import cn from "classnames";

import "./styles.scss";

export type TableProps = ComponentPropsWithoutRef<"table">;
export const Table = forwardRef<HTMLTableElement, TableProps>(function Thead(
  { className, ...props },
  ref,
) {
  return (
    <table ref={ref} className={cn("govuk-table", "lbh-table", className)} {...props} />
  );
});

export type TheadProps = ComponentPropsWithoutRef<"thead">;
export const Thead = forwardRef<HTMLTableSectionElement, TheadProps>(function Thead(
  { className, ...props },
  ref,
) {
  return <thead ref={ref} className={cn("govuk-table__head", className)} {...props} />;
});

export type TbodyProps = ComponentPropsWithoutRef<"tbody">;
export const Tbody = forwardRef<HTMLTableSectionElement, TbodyProps>(function Tbody(
  { className, ...props },
  ref,
) {
  return <tbody ref={ref} className={cn("govuk-table__body", className)} {...props} />;
});

export type TrProps = ComponentPropsWithoutRef<"tr">;
export const Tr = forwardRef<HTMLTableRowElement, TrProps>(function Tr(
  { className, ...props },
  ref,
) {
  return <tr ref={ref} className={cn("govuk-table__row", className)} {...props} />;
});

export interface ThProps extends ComponentPropsWithoutRef<"th"> {
  isNumeric?: boolean;
}
export const Th = forwardRef<HTMLTableHeaderCellElement, ThProps>(function Th(
  { className, isNumeric, ...props },
  ref,
) {
  const thClasses = cn(
    "govuk-table__cell",
    "govuk-table__header",
    {
      "govuk-table__cell--numeric": isNumeric,
    },
    className,
  );

  return <th ref={ref} className={thClasses} {...props} />;
});

export interface TdProps extends ComponentPropsWithoutRef<"td"> {
  isNumeric?: boolean;
}
export const Td = forwardRef<HTMLTableCellElement, TdProps>(function Td(
  { className, isNumeric, ...props },
  ref,
) {
  const tdClasses = cn(
    "govuk-table__cell",
    {
      "govuk-table__cell--numeric": isNumeric,
    },
    className,
  );

  return <td ref={ref} className={tdClasses} {...props} />;
});

export interface TableCaptionProps extends ComponentPropsWithoutRef<"caption"> {
  variant?: "xlarge" | "large" | "medium" | "small";
}
export const TableCaption = forwardRef<HTMLTableCaptionElement, TableCaptionProps>(
  function TableCaption({ children, className, variant, ...props }, ref) {
    const captionClasses = cn(
      "govuk-table__caption",
      "lbh-table__caption",
      {
        "govuk-table__caption--s": variant === "small",
        "govuk-table__caption--m": variant === "medium",
        "govuk-table__caption--l": variant === "large",
        "govuk-table__caption--xl": variant === "xlarge",
      },
      className,
    );

    return (
      <caption ref={ref} className={captionClasses} {...props}>
        {children}
      </caption>
    );
  },
);
