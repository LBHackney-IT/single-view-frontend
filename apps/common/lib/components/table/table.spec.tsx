import React from "react";

import { render, testA11y } from "@hackney/mtfh-test-utils";

import { Table, TableCaption, Tbody, Td, Th, Thead, Tr } from "./table";

const TableHeadSample = () => (
  <Thead>
    <Tr>
      <Th>Month you apply</Th>
      <Th>Rate for bicycles</Th>
      <Th>Rate for vehicles</Th>
    </Tr>
  </Thead>
);

const TableBodySample = () => (
  <Tbody>
    <Tr>
      <Td>January</Td>
      <Td isNumeric>£85</Td>
      <Td isNumeric>£95</Td>
    </Tr>
    <Tr>
      <Td>February</Td>
      <Td isNumeric>£75</Td>
      <Td isNumeric>£55</Td>
    </Tr>
    <Tr>
      <Td>March</Td>
      <Td isNumeric>£165</Td>
      <Td isNumeric>£125</Td>
    </Tr>
  </Tbody>
);

test("it displays a basic table passes a11y", async () => {
  const { container } = render(
    <Table>
      <TableBodySample />
    </Table>,
  );
  expect(container).toMatchSnapshot();
  await testA11y(container);
});

test("it displays a custom styled basic table passes a11y", async () => {
  const tableClass = "customTableClass";
  const { container } = render(
    <Table className={tableClass}>
      <TableBodySample />
    </Table>,
  );
  expect(container).toMatchSnapshot();
  await testA11y(container);
});

test("it displays a table with a header and passes a11y", async () => {
  const { container } = render(
    <Table>
      <TableHeadSample />
      <TableBodySample />
    </Table>,
  );
  expect(container).toMatchSnapshot();
  await testA11y(container);
});

test("it displays a table with a caption and header and passes a11y", async () => {
  const { container } = render(
    <Table>
      <TableCaption>Caption 1: Months and rates</TableCaption>
      <TableHeadSample />
      <TableBodySample />
    </Table>,
  );
  expect(container).toMatchSnapshot();
  await testA11y(container);
});

const captionSizes = ["xlarge", "large", "medium", "small"];
captionSizes.map((size) =>
  test(`it displays a table with a ${size} caption and header and passes a11y`, async () => {
    const { container } = render(
      <Table>
        <TableCaption>{size} Caption 1: Months and rates</TableCaption>
        <TableHeadSample />
        <TableBodySample />
      </Table>,
    );
    expect(container).toMatchSnapshot();
    await testA11y(container);
  }),
);
