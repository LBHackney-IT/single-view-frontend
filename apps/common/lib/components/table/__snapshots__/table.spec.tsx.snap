// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it displays a basic table passes a11y 1`] = `
<div>
  <table
    class="govuk-table lbh-table"
  >
    <tbody
      class="govuk-table__body"
    >
      <tr
        class="govuk-table__row"
      >
        <td
          class="govuk-table__cell"
        >
          January
        </td>
        <td
          class="govuk-table__cell govuk-table__cell--numeric"
        >
          £85
        </td>
        <td
          class="govuk-table__cell govuk-table__cell--numeric"
        >
          £95
        </td>
      </tr>
      <tr
        class="govuk-table__row"
      >
        <td
          class="govuk-table__cell"
        >
          February
        </td>
        <td
          class="govuk-table__cell govuk-table__cell--numeric"
        >
          £75
        </td>
        <td
          class="govuk-table__cell govuk-table__cell--numeric"
        >
          £55
        </td>
      </tr>
      <tr
        class="govuk-table__row"
      >
        <td
          class="govuk-table__cell"
        >
          March
        </td>
        <td
          class="govuk-table__cell govuk-table__cell--numeric"
        >
          £165
        </td>
        <td
          class="govuk-table__cell govuk-table__cell--numeric"
        >
          £125
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`it displays a custom styled basic table passes a11y 1`] = `
<div>
  <table
    class="govuk-table lbh-table customTableClass"
  >
    <tbody
      class="govuk-table__body"
    >
      <tr
        class="govuk-table__row"
      >
        <td
          class="govuk-table__cell"
        >
          January
        </td>
        <td
          class="govuk-table__cell govuk-table__cell--numeric"
        >
          £85
        </td>
        <td
          class="govuk-table__cell govuk-table__cell--numeric"
        >
          £95
        </td>
      </tr>
      <tr
        class="govuk-table__row"
      >
        <td
          class="govuk-table__cell"
        >
          February
        </td>
        <td
          class="govuk-table__cell govuk-table__cell--numeric"
        >
          £75
        </td>
        <td
          class="govuk-table__cell govuk-table__cell--numeric"
        >
          £55
        </td>
      </tr>
      <tr
        class="govuk-table__row"
      >
        <td
          class="govuk-table__cell"
        >
          March
        </td>
        <td
          class="govuk-table__cell govuk-table__cell--numeric"
        >
          £165
        </td>
        <td
          class="govuk-table__cell govuk-table__cell--numeric"
        >
          £125
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`it displays a table with a caption and header and passes a11y 1`] = `
<div>
  <table
    class="govuk-table lbh-table"
  >
    <caption
      class="govuk-table__caption lbh-table__caption"
    >
      Caption 1: Months and rates
    </caption>
    <thead
      class="govuk-table__head"
    >
      <tr
        class="govuk-table__row"
      >
        <th
          class="govuk-table__cell govuk-table__header"
        >
          Month you apply
        </th>
        <th
          class="govuk-table__cell govuk-table__header"
        >
          Rate for bicycles
        </th>
        <th
          class="govuk-table__cell govuk-table__header"
        >
          Rate for vehicles
        </th>
      </tr>
    </thead>
    <tbody
      class="govuk-table__body"
    >
      <tr
        class="govuk-table__row"
      >
        <td
          class="govuk-table__cell"
        >
          January
        </td>
        <td
          class="govuk-table__cell govuk-table__cell--numeric"
        >
          £85
        </td>
        <td
          class="govuk-table__cell govuk-table__cell--numeric"
        >
          £95
        </td>
      </tr>
      <tr
        class="govuk-table__row"
      >
        <td
          class="govuk-table__cell"
        >
          February
        </td>
        <td
          class="govuk-table__cell govuk-table__cell--numeric"
        >
          £75
        </td>
        <td
          class="govuk-table__cell govuk-table__cell--numeric"
        >
          £55
        </td>
      </tr>
      <tr
        class="govuk-table__row"
      >
        <td
          class="govuk-table__cell"
        >
          March
        </td>
        <td
          class="govuk-table__cell govuk-table__cell--numeric"
        >
          £165
        </td>
        <td
          class="govuk-table__cell govuk-table__cell--numeric"
        >
          £125
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`it displays a table with a header and passes a11y 1`] = `
<div>
  <table
    class="govuk-table lbh-table"
  >
    <thead
      class="govuk-table__head"
    >
      <tr
        class="govuk-table__row"
      >
        <th
          class="govuk-table__cell govuk-table__header"
        >
          Month you apply
        </th>
        <th
          class="govuk-table__cell govuk-table__header"
        >
          Rate for bicycles
        </th>
        <th
          class="govuk-table__cell govuk-table__header"
        >
          Rate for vehicles
        </th>
      </tr>
    </thead>
    <tbody
      class="govuk-table__body"
    >
      <tr
        class="govuk-table__row"
      >
        <td
          class="govuk-table__cell"
        >
          January
        </td>
        <td
          class="govuk-table__cell govuk-table__cell--numeric"
        >
          £85
        </td>
        <td
          class="govuk-table__cell govuk-table__cell--numeric"
        >
          £95
        </td>
      </tr>
      <tr
        class="govuk-table__row"
      >
        <td
          class="govuk-table__cell"
        >
          February
        </td>
        <td
          class="govuk-table__cell govuk-table__cell--numeric"
        >
          £75
        </td>
        <td
          class="govuk-table__cell govuk-table__cell--numeric"
        >
          £55
        </td>
      </tr>
      <tr
        class="govuk-table__row"
      >
        <td
          class="govuk-table__cell"
        >
          March
        </td>
        <td
          class="govuk-table__cell govuk-table__cell--numeric"
        >
          £165
        </td>
        <td
          class="govuk-table__cell govuk-table__cell--numeric"
        >
          £125
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`it displays a table with a large caption and header and passes a11y 1`] = `
<div>
  <table
    class="govuk-table lbh-table"
  >
    <caption
      class="govuk-table__caption lbh-table__caption"
    >
      large
       Caption 1: Months and rates
    </caption>
    <thead
      class="govuk-table__head"
    >
      <tr
        class="govuk-table__row"
      >
        <th
          class="govuk-table__cell govuk-table__header"
        >
          Month you apply
        </th>
        <th
          class="govuk-table__cell govuk-table__header"
        >
          Rate for bicycles
        </th>
        <th
          class="govuk-table__cell govuk-table__header"
        >
          Rate for vehicles
        </th>
      </tr>
    </thead>
    <tbody
      class="govuk-table__body"
    >
      <tr
        class="govuk-table__row"
      >
        <td
          class="govuk-table__cell"
        >
          January
        </td>
        <td
          class="govuk-table__cell govuk-table__cell--numeric"
        >
          £85
        </td>
        <td
          class="govuk-table__cell govuk-table__cell--numeric"
        >
          £95
        </td>
      </tr>
      <tr
        class="govuk-table__row"
      >
        <td
          class="govuk-table__cell"
        >
          February
        </td>
        <td
          class="govuk-table__cell govuk-table__cell--numeric"
        >
          £75
        </td>
        <td
          class="govuk-table__cell govuk-table__cell--numeric"
        >
          £55
        </td>
      </tr>
      <tr
        class="govuk-table__row"
      >
        <td
          class="govuk-table__cell"
        >
          March
        </td>
        <td
          class="govuk-table__cell govuk-table__cell--numeric"
        >
          £165
        </td>
        <td
          class="govuk-table__cell govuk-table__cell--numeric"
        >
          £125
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`it displays a table with a medium caption and header and passes a11y 1`] = `
<div>
  <table
    class="govuk-table lbh-table"
  >
    <caption
      class="govuk-table__caption lbh-table__caption"
    >
      medium
       Caption 1: Months and rates
    </caption>
    <thead
      class="govuk-table__head"
    >
      <tr
        class="govuk-table__row"
      >
        <th
          class="govuk-table__cell govuk-table__header"
        >
          Month you apply
        </th>
        <th
          class="govuk-table__cell govuk-table__header"
        >
          Rate for bicycles
        </th>
        <th
          class="govuk-table__cell govuk-table__header"
        >
          Rate for vehicles
        </th>
      </tr>
    </thead>
    <tbody
      class="govuk-table__body"
    >
      <tr
        class="govuk-table__row"
      >
        <td
          class="govuk-table__cell"
        >
          January
        </td>
        <td
          class="govuk-table__cell govuk-table__cell--numeric"
        >
          £85
        </td>
        <td
          class="govuk-table__cell govuk-table__cell--numeric"
        >
          £95
        </td>
      </tr>
      <tr
        class="govuk-table__row"
      >
        <td
          class="govuk-table__cell"
        >
          February
        </td>
        <td
          class="govuk-table__cell govuk-table__cell--numeric"
        >
          £75
        </td>
        <td
          class="govuk-table__cell govuk-table__cell--numeric"
        >
          £55
        </td>
      </tr>
      <tr
        class="govuk-table__row"
      >
        <td
          class="govuk-table__cell"
        >
          March
        </td>
        <td
          class="govuk-table__cell govuk-table__cell--numeric"
        >
          £165
        </td>
        <td
          class="govuk-table__cell govuk-table__cell--numeric"
        >
          £125
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`it displays a table with a small caption and header and passes a11y 1`] = `
<div>
  <table
    class="govuk-table lbh-table"
  >
    <caption
      class="govuk-table__caption lbh-table__caption"
    >
      small
       Caption 1: Months and rates
    </caption>
    <thead
      class="govuk-table__head"
    >
      <tr
        class="govuk-table__row"
      >
        <th
          class="govuk-table__cell govuk-table__header"
        >
          Month you apply
        </th>
        <th
          class="govuk-table__cell govuk-table__header"
        >
          Rate for bicycles
        </th>
        <th
          class="govuk-table__cell govuk-table__header"
        >
          Rate for vehicles
        </th>
      </tr>
    </thead>
    <tbody
      class="govuk-table__body"
    >
      <tr
        class="govuk-table__row"
      >
        <td
          class="govuk-table__cell"
        >
          January
        </td>
        <td
          class="govuk-table__cell govuk-table__cell--numeric"
        >
          £85
        </td>
        <td
          class="govuk-table__cell govuk-table__cell--numeric"
        >
          £95
        </td>
      </tr>
      <tr
        class="govuk-table__row"
      >
        <td
          class="govuk-table__cell"
        >
          February
        </td>
        <td
          class="govuk-table__cell govuk-table__cell--numeric"
        >
          £75
        </td>
        <td
          class="govuk-table__cell govuk-table__cell--numeric"
        >
          £55
        </td>
      </tr>
      <tr
        class="govuk-table__row"
      >
        <td
          class="govuk-table__cell"
        >
          March
        </td>
        <td
          class="govuk-table__cell govuk-table__cell--numeric"
        >
          £165
        </td>
        <td
          class="govuk-table__cell govuk-table__cell--numeric"
        >
          £125
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`it displays a table with a xlarge caption and header and passes a11y 1`] = `
<div>
  <table
    class="govuk-table lbh-table"
  >
    <caption
      class="govuk-table__caption lbh-table__caption"
    >
      xlarge
       Caption 1: Months and rates
    </caption>
    <thead
      class="govuk-table__head"
    >
      <tr
        class="govuk-table__row"
      >
        <th
          class="govuk-table__cell govuk-table__header"
        >
          Month you apply
        </th>
        <th
          class="govuk-table__cell govuk-table__header"
        >
          Rate for bicycles
        </th>
        <th
          class="govuk-table__cell govuk-table__header"
        >
          Rate for vehicles
        </th>
      </tr>
    </thead>
    <tbody
      class="govuk-table__body"
    >
      <tr
        class="govuk-table__row"
      >
        <td
          class="govuk-table__cell"
        >
          January
        </td>
        <td
          class="govuk-table__cell govuk-table__cell--numeric"
        >
          £85
        </td>
        <td
          class="govuk-table__cell govuk-table__cell--numeric"
        >
          £95
        </td>
      </tr>
      <tr
        class="govuk-table__row"
      >
        <td
          class="govuk-table__cell"
        >
          February
        </td>
        <td
          class="govuk-table__cell govuk-table__cell--numeric"
        >
          £75
        </td>
        <td
          class="govuk-table__cell govuk-table__cell--numeric"
        >
          £55
        </td>
      </tr>
      <tr
        class="govuk-table__row"
      >
        <td
          class="govuk-table__cell"
        >
          March
        </td>
        <td
          class="govuk-table__cell govuk-table__cell--numeric"
        >
          £165
        </td>
        <td
          class="govuk-table__cell govuk-table__cell--numeric"
        >
          £125
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
